tests/test_migrations.py

```diff
@@ -151,6 +151,86 @@ def test_pre_migration_modifies_answers(
+        build_file_tree(
+            {
+                "[[ _copier_conf.answers_file ]].jinja": (
+                    "[[ _copier_answers|tojson ]]"
+                ),
+                "copier.yml": (
+                    f"""\
+                    _envops: {BRACKET_ENVOPS_JSON}
+                    best_song: la vie en rose
+                    """
+                ),
+                "songs.json.jinja": "[ [[ best_song|tojson ]] ]",
+            }
+        )
+        git("init")
+        git("add", ".")
+        git("commit", "-m1")
+        git("tag", "v1")
+    # User copies v1 template into subproject
+    with local.cwd(dst):
+        run_copy(src_path=str(src), defaults=True, overwrite=True)
+        answers = json.loads(Path(".copier-answers.yml").read_text())
+        assert answers["_commit"] == "v1"
+        assert answers["best_song"] == "la vie en rose"
+        assert json.loads(Path("songs.json").read_text()) == ["la vie en rose"]
+        git("init")
+        git("add", ".")
+        git("commit", "-m1")
+    with local.cwd(src):
+        build_file_tree(
+            {
+                # v2 of template supports multiple songs, has a different default
+                # and includes a data format migration script
+                "copier.yml": (
+                    f"""\
+                    _envops: {BRACKET_ENVOPS_JSON}
+                    best_song_list:
+                        default: [paranoid android]
+                    bad_song:
+                        default: [creep]
+                    _migrations:
+                    -   version: v2
+                        before:
+                        -   - python
+                            - -c
+                            - |
+                                import sys, json, pathlib
+                                answers_path = pathlib.Path(*sys.argv[1:])
+                                answers = json.loads(answers_path.read_text())
+                                answers["best_song_list"] = [answers.pop("best_song")]
+                                answers_path.write_text(json.dumps(answers))
+                            - "[[ _copier_conf.dst_path ]]"
+                            - "[[ _copier_conf.answers_file ]]"
+                    """
+                ),
+                "songs.json.jinja": "[[ best_song_list|tojson ]]",
+            }
+        )
+        git("add", ".")
+        git("commit", "-m2")
+        git("tag", "v2")
+    # User updates subproject to v2 template
+    with local.cwd(dst):
+        run_update(defaults=True, overwrite=True, unsafe=True, skip_answered=True)
+        answers = json.loads(Path(".copier-answers.yml").read_text())
+        assert answers["_commit"] == "v2"
+        assert "best_song" not in answers
+        assert answers["best_song_list"] == ["paranoid android"]
+        assert json.loads(Path("songs.json").read_text()) == ["paranoid android"]
+        assert answers["bad_song"] == ["creep"]
```

Comment:
I already hand-tested this feature. The feature is working fine, but the test isn't appropriate.

What this test is testing is that the migration is correct. However, this feature is not about migrations, but about interactive questions.

In v1 of template you create in the test, you add the question `best_song`. In v2, you replace that one with `bad_song` and `best_song_list`. Since the old question disappeared in v2, it wouldn't be answered, even without this feature.

Your feature is only noticed when using Copier interactively, but you're adding `defaults=True, overwrite=True` and disabling interactiveness.

So, summarizing, you can remove this test and add a new one that:
- Is interactive.
- Checks that the old question isn't asked again when updating.
- Checks that new questions are asked.

Interactive (TUI) tests are a bit more complex, so you can inspire yourself by this other test: https://github.com/copier-org/copier/blob/d6c04b478d5a4126dc10181e884ecd838d79f436/tests/test_prompt.py#L49-L117

Reply:
Hey, yeah, it makes sense. 
I've removed my old test and added 2 more. 

Can you check it?

Reply:
Much better now, thanks!